### PR TITLE
Remove `rustc_session::config::rustc_short_optgroups`

### PR DESCRIPTION
**Pull request by Zalathar**
Follow-up to https://github.com/rust-lang/rust/pull/132754#discussion_r1835427349.

The name `rustc_short_optgroups` has always been confusing, because it is unrelated to the distinction between short and long options (i.e. `-s` vs `--long`), and instead means something like “the subset of command-line options that are printed by `rustc --help` without `-v`”.

So let's merge that function into the main `rustc_optgroups`, and store the relevant bit of information in a boolean field in `RustcOptGroup` instead.

---

This PR also modifies `RustcOptGroup` to store its various strings directly, instead of inside a boxed `apply` closure. That turned out to not be necessary for the main change, but is a worthwhile cleanup in its own right.

### CODE REVIEW

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1550,21 +1557,11 @@ pub fn rustc_short_optgroups() -> Vec<RustcOptGroup> {
...
-/// long-term interface for rustc.
-pub fn rustc_optgroups() -> Vec<RustcOptGroup> {
-    use OptionKind::{Multi, Opt};
-    use OptionStability::{Stable, Unstable};
-
-    use self::make_opt as opt;
+    ];
 
-    let mut opts = rustc_short_optgroups();
-    // FIXME: none of these descriptions are actually used
```
Comment (Zalathar) [inline compiler/rustc_session/src/config.rs:1566]: This comment was incorrect; the descriptions are used by `rustc --help -v`, as demonstrated by `tests/run-make/rustc-help`.

**Comment by jieyouxu** [comment]
r? jieyouxu

File: compiler/rustc_session/src/config.rs [hunk 2]
```diff
@@ -1398,9 +1398,16 @@ pub enum OptionKind {
 }
 
 pub struct RustcOptGroup {
-    apply: Box<dyn Fn(&mut getopts::Options) -> &mut getopts::Options>,
+    /// Display name for this option. Normally equal to `long_name`, except for
+    /// options that don't have a long name.
     pub name: &'static str,
```
Comment (jieyouxu) [inline compiler/rustc_session/src/config.rs:1403]: Suggestion: `display_name`? We have `short_name` and `long_name` below, so having just `name` feels a bit confusing.

File: compiler/rustc_session/src/config.rs [hunk 3]
```diff
@@ -1408,6 +1408,10 @@ pub struct RustcOptGroup {
     long_name: &'static str,
     desc: &'static str,
     value_hint: &'static str,
+
+    /// If true, this option should not be printed by `rustc --help`, but
+    /// should still be printed by `rustc --help -v`.
+    pub is_verbose_help_only: bool,
```
Comment (jieyouxu) [inline compiler/rustc_session/src/config.rs:1414]: Remark: yes thank you, this is so much clearer.

File: compiler/rustc_session/src/config.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_session/src/config.rs:1566]: lol

**Review by jieyouxu — APPROVED** [review]
Very nice cleanup! One minor naming nit then r=me.

File: compiler/rustc_session/src/config.rs (hunk 2 above)
Comment (Zalathar) [reply compiler/rustc_session/src/config.rs:1403]: So, in the process of looking into this, I remembered that things are actually more complicated.

We do use this as a display name, but it also gets used for other things, such as matching against the option name reported by some `getopts` errors.

So I'm going to push a better comment, but now I don't like the idea of trying to rename `name` in this PR.

**Comment by Zalathar** [comment]
Revised comment for `name` [(diff)](https://github.com/rust-lang/rust/compare/e4d6ff5796933c667afa8986f8aa8b56463b8396..8b4701d74cc7bc1a9532716fd2bd25579102115d), in light of the fact that it's not *just* a display name.

@rustbot ready

File: compiler/rustc_session/src/config.rs (hunk 2 above)
Comment (jieyouxu) [reply compiler/rustc_session/src/config.rs]: Yes, that's fine with me.

**Review by jieyouxu — APPROVED** [review]
Cool

**Pull request merged**